File: sentimental-credit/credit.py
import cs50


def main():
    num = cs50.get_int("Number: ")

    num_len = len(str(num))
    first_2_digits = int(str(num)[:2])

    if not valid_card_num(num):
        print("INVALID")
        return 1

    # American Express
    if num_len == 15 and (first_2_digits == 34 or first_2_digits == 37):
        print("AMEX")
    # Mastercard
    elif num_len == 16 and first_2_digits >= 51 and first_2_digits <= 55:
        print("MASTERCARD")
    # VISA
    elif (num_len == 13 or num_len == 16) and int(first_2_digits / 10) == 4:
        print("VISA")
    else:
        print("INVALID")


def valid_card_num(num):
    sum = 0
    # num as a string
    n = str(num)
    # Tracks every first and secnod digit
    tracker = 0

    # Iterate through every digit
    for i in reversed(n):
        digit = int(i)
        if tracker % 2 != 0:
            product = digit * 2
            # If product is 2 digits or not
            sum += product % 10 + int(product / 10)
        else:
            sum += digit

        tracker += 1

    return sum % 10 == 0


if __name__ == "__main__":
    main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

bool cardNumIsValid(long num);

int main(void)
{
    // Ask for credit card #
    long cardNum = get_long("Enter a credit card number: ");

    // Length
    long cardNumLength = 0;
    for (long i = cardNum; i > 0; i /= 10)
    {
        ++cardNumLength;
    }

    // First 2 digits
    long first2Digits = cardNum;
    while (first2Digits > 99)
    {
        first2Digits /= 10;
    }

    if (cardNumIsValid(cardNum))
    {
        // American Express
        if (cardNumLength == 15 && (first2Digits == 34 || first2Digits == 37))
        {
            printf("AMEX\n");
        }
        // MasterCard
        else if (cardNumLength == 16 && first2Digits >= 51 && first2Digits <= 55)
        {
            printf("MASTERCARD\n");
        }
        // Visa
        else if ((cardNumLength == 13 || cardNumLength == 16) && first2Digits / 10 == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

}

// Return if the credit card number is valid
bool cardNumIsValid(long num)
{
    long sum = 0;
    for (long i = num / 10; i > 0; i /= 100)
    {
        long product = i % 10 * 2;
        // If product is 2 digits
        if (product > 9)
        {
            sum += product % 10 + product / 10;
        }
        else
        {
            sum += product;
        }
    }
    for (long i = num; i > 0; i /= 100)
    {
        long digit = i % 10;
        sum += digit;
    }
    return sum % 10 == 0;
}